design-proposals/klip-50-partition-and-offset-in-ksqldb.md

```diff
@@ -0,0 +1,253 @@
+is semantically dubious, especially for tables that are the result of other
+computation such as an aggregation over a stream. Should `ROWPARTITION` and `ROWOFFSET`
+in such a case represent the partition and offset of the message in the source stream
+that resulted in the aggregate value, or should they represent the partition and offset
+of the aggregate value in the output table? Note that timestamps don't have this
+ambiguity as the timestamp of the two records mentioned are the same.
+
+The second reason has to do with implications on state store size. ksqlDB is able to 
+support `ROWTIME` in pull queries because the state stores from which pull queries are 
+served includes the timestamp for the latest table row (for each key). In contrast, 
+message partition and offset are not saved in the state stores. As a result, supporting 
+`ROWPARTITION` and `ROWOFFSET` in pull queries would require adding this information
+to state stores. (I think there might be a way to get the partition by other means,
+from the replica queried in order to serve the pull query, but offset information
+definitely wouldn't be available.)
+
+If a user really wants to make partition and/or offset information available for pull
+queries, they can choose to explicitly include the source partition and offset as
+part of the query that materializes their table and query for partition/offset as
+value columns as part of their pull query.
+
+### Other semantics
+
+Similar to `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will not 
+be included in the list of columns returned as part of a source description.
+
+`ROWPARTITION` and `ROWOFFSET` will also not be added to the output of `PRINT TOPIC` commands.
+
+Unlike `ROWTIME`, the new `ROWPARTITION` and `ROWOFFSET` pseudocolumns will not
+be allowed as part of `INSERT VALUES` statements. It does not make sense for users
+to specify the offset, and while it's technically possible for users to specify a
+custom partition, this does not seem useful as violating expected partitioning has
+implications for downstream queries.
+
+## Design
+
+`ROWPARTITION` and `ROWOFFSET` will be added as new pseudocolumns, similar to how
+`ROWTIME` is already implemented today. New logic will be added to disallow the use of
+`ROWPARTITION` and `ROWOFFSET` in pull queries, while `ROWTIME` will still be allowed.
+
+However, directly adding `ROWPARTITION` and `ROWOFFSET` as new pseudocolumns would break 
+compatibility for certain existing queries since, in some situations, 
+[ksqlDB writes pseudocolumns into repartition and changelog topics](https://github.com/confluentinc/ksql/issues/7489). 
+This is problematic because adding new pseudocolumns would change the schema of these 
+intermediate topics (and state stores) and break running queries as a result.
+
+In light of the above, we have two approaches for implementation:
+1. Update ksqlDB to never write pseudocolumns into repartition and changelog topics 
```

Comment:
I thought the ultimate solution would be https://github.com/confluentinc/ksql/issues/6374. More specifically, since the partition/offset are only for the sink topics, so conceptually intermediate topics would never need to encode the partition/offset at all assuming we have clear separation between physical and logical schemas, no?

Reply:
> since the partition/offset are only for the sink topics

As discussed above, the intention was for `ROWPARTITION` and `ROWOFFSET` to represent the source partition and offset, rather than the sink partition and offset. Let's resolve that discussion first, and return to this if it's still relevant.

Reply:
When punctuation is triggered, we may send records to downstream which do not have partition/offset/etc metadata since they are not triggered by data, but by time. Which means, we may need to be able to set some sentinel values in the corresponding columns for records generated by punctuations.

Yes vector clocks are for the context of punctuations, they can cover both data-driven and time-driven (a.k.a. punctuation) events, but are flushed across chains of query topologies with the source topic metadata (partition/offset) in it. E.g. say we have a chain of queries:

```
source -> query1 -> topic1 -> query2 -> topic2 -> query3 -> sink
```

Then the metadata would always be representing the `source` topic, flowing through topic1/2 and sink. For this KLIP we may need the metadata to be representing the `source`, `topic1`, and `topic2` respectively for query1/2/3, so for now we can just use the context metadata to achieve this. I'm just thinking time-driven results maybe we can do piggyback on the same mechanism (a.k.a. control messages) to do that in the longer term.

Reply:
ksqlDB and KS DSL don't use punctuations -- so we should not worry about it? -- I also think that vector clocks won't help, as we need the partition/offset information on a per record basis while vector-clock data would only be written periodically.